File: sdks/cpp/examples/motion_automation.cpp
#include <ora/client.hpp>
#include <iostream>
#include <thread>
#include <chrono>

class MotionLightAutomation {
public:
    MotionLightAutomation(ora::Client& client, const std::string& motion_sensor,
                         const std::string& light_entity, int brightness, int timeout)
        : client_(client), motion_sensor_(motion_sensor),
          light_entity_(light_entity), brightness_(brightness), timeout_(timeout) {}

    void on_motion_detected() {
        std::cout << "Motion detected!\n";

        // Turn on light
        nlohmann::json data = {{"brightness", brightness_}};
        client_.turn_on(light_entity_, data);

        // Send notification
        ora::NotificationPayload notification{
            .title = "Motion Detected",
            .message = "Turned on " + light_entity_ + " at " + std::to_string(brightness_) + " brightness",
            .priority = "normal"
        };
        client_.send_notification(notification);

        // Store activation time
        auto now = std::chrono::system_clock::now();
        auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(
            now.time_since_epoch()
        ).count();

        client_.set_storage("last_activation", {{"timestamp", timestamp}});
    }

    void on_motion_cleared() {
        std::cout << "Motion cleared, waiting " << timeout_ << "s...\n";

        // Wait for timeout
        std::this_thread::sleep_for(std::chrono::seconds(timeout_));

        // Check if motion is still off
        auto sensor = client_.get_entity(motion_sensor_);
        if (sensor.state == "off") {
            client_.turn_off(light_entity_);

            ora::NotificationPayload notification{
                .title = "Auto-Off",
                .message = "Turned off " + light_entity_ + " after " + std::to_string(timeout_) + "s",
                .priority = "normal"
            };
            client_.send_notification(notification);

            std::cout << "Light turned off\n";
        }
    }

private:
    ora::Client& client_;
    std::string motion_sensor_;
    std::string light_entity_;
    int brightness_;
    int timeout_;
};

int main() {
    try {
        ora::Client client("http://localhost:8080", "api-key");

        MotionLightAutomation automation(
            client,
            "binary_sensor.motion_living_room",
            "light.living_room",
            255,  // brightness
            60    // timeout in seconds
        );

        // Simulate motion detection
        std::cout << "Simulating motion automation...\n";
        automation.on_motion_detected();

        // Wait and simulate motion clearing
        std::this_thread::sleep_for(std::chrono::seconds(2));
        automation.on_motion_cleared();

        std::cout << "Automation complete!\n";

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: sdks/python/examples/motion_automation.py
"""
Example: Motion-activated light automation
"""

from rumahl_sdk import AutomationPlugin, PluginContext
from typing import Dict, Any


class MotionLightAutomation(AutomationPlugin):
    """
    Turn on lights when motion is detected

    Settings:
    - light_entity_id: Entity ID of light to control
    - motion_sensor_id: Entity ID of motion sensor
    - brightness: Brightness level (0-255)
    - timeout: Auto-off timeout in seconds
    """

    async def on_event(self, context: PluginContext, event: Dict[str, Any]) -> None:
        # Check if this is a state change event
        if event.get("type") != "entity_state_changed":
            return

        entity_id = event["data"]["entity_id"]
        new_state = event["data"]["new_state"]

        # Get configured motion sensor
        motion_sensor_id = context.settings.get("motion_sensor_id")
        if entity_id != motion_sensor_id:
            return

        # Get light settings
        light_id = context.settings.get("light_entity_id")
        brightness = context.settings.get("brightness", 255)

        # Turn on light when motion detected
        if new_state == "on":
            await context.call_service("light", "turn_on", light_id, {"brightness": brightness})

            await context.notify(
                "Motion Detected", f"Turned on {light_id} at {brightness} brightness"
            )

            # Store activation time
            import time

            await context.store("last_activation", time.time())

        # Turn off light when motion clears
        elif new_state == "off":
            timeout = context.settings.get("timeout", 60)

            # Wait for timeout
            import asyncio

            await asyncio.sleep(timeout)

            # Check if motion is still off
            sensor = await context.get_entity(motion_sensor_id)
            if sensor.state == "off":
                await context.call_service("light", "turn_off", light_id)
                await context.notify("Auto-Off", f"Turned off {light_id} after {timeout}s")
